File: exercise/python/day13/model/christmas_elf.py
from .offer import Offer
from .receipt import Receipt

class ChristmasElf:
    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    def add_special_offer(self, offer_type, product, argument):
        self.offers[product] = Offer(offer_type, product, argument)

    def checks_out_articles_from(self, sleigh):
        receipt = Receipt()

        for product, quantity in sleigh.get_items():
            unit_price = self.catalog.get_unit_price(product)
            total_price = quantity * unit_price
            receipt.add_product(product, quantity, unit_price, total_price)

        sleigh.handle_offers(receipt, self.offers, self.catalog)

        return receipt

File: exercise/ts/day13/src/santamarket.model/ChristmasElf.ts
import { SantamarketCatalog } from './SantamarketCatalog';
import { Product } from './Product';
import { Offer } from './Offer';
import { SpecialOfferType } from './SpecialOfferType';
import { ShoppingSleigh } from './ShoppingSleigh';
import { Receipt } from './Receipt';

export class ChristmasElf {
    private readonly catalog: SantamarketCatalog;
    private readonly offers: Map<Product, Offer>;

    constructor(catalog: SantamarketCatalog) {
        this.catalog = catalog;
        this.offers = new Map<Product, Offer>();
    }

    addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: number): void {
        this.offers.set(product, new Offer(offerType, product, argument));
    }

    checksOutArticlesFrom(sleigh: ShoppingSleigh): Receipt {
        const receipt = new Receipt();
        const productQuantities = sleigh.getItems();

        productQuantities.forEach(({ product, quantity }) => {
            const unitPrice = this.catalog.getUnitPrice(product);
            const totalPrice = quantity * unitPrice;
            receipt.addProduct(product, quantity, unitPrice, totalPrice);
        });

        sleigh.handleOffers(receipt, this.offers, this.catalog);

        return receipt;
    }
}
